peer/brontide.go

```diff
@@ -3648,18 +3643,52 @@ func (p *Brontide) handleCloseMsg(msg *closeMsg) {
 			chanCloser.CloseRequest().Err <- err
 		}
 		delete(p.activeChanCloses, msg.cid)
-		return
+
+		p.Disconnect(err)
 	}
 
-	// Queue any messages to the remote peer that need to be sent as a part of
-	// this latest round of negotiations.
-	for _, msg := range msgs {
-		p.queueMsg(msg, nil)
+	// Next, we'll process the next message using the target state machine.
+	// We'll either continue negotiation, or halt.
+	switch typed := msg.msg.(type) {
+	case *lnwire.Shutdown:
+		oShutdown, err := chanCloser.ReceiveShutdown(*typed)
+		if err != nil {
+			handleErr(err)
+			return
+		}
+
+		oShutdown.WhenSome(func(msg lnwire.Shutdown) {
+			p.queueMsg(&msg, nil)
+		})
+
+		oClosingSigned, err := chanCloser.BeginNegotiation()
+		if err != nil {
+			handleErr(err)
+			return
+		}
+
+		oClosingSigned.WhenSome(func(msg lnwire.ClosingSigned) {
+			p.queueMsg(&msg, nil)
+		})
+
+	case *lnwire.ClosingSigned:
+		oClosingSigned, err := chanCloser.ReceiveClosingSigned(*typed)
+		if err != nil {
+			handleErr(err)
+			return
+		}
+
+		oClosingSigned.WhenSome(func(msg lnwire.ClosingSigned) {
+			p.queueMsg(&msg, nil)
+		})
+
+	default:
+		panic("impossible closeMsg type")
```

Comment:
I did this to make it very clear if we implement a "closeMsg" type that we don't handle in this switch case. I actually think the aggressiveness of this blowup will reduce production vulnerabilities not increase them. Things that should be **structurally** impossible should have aggressive blowups so we catch them immediately. Crashes are the most obvious failures and the ones least likely to slip by our testing. If we do something other than panic, then the "impossibility" (read: core assumption we make about the system) silently breaks down and causes other bugs. Every bug has the potential to be a security issue and so doing something less aggressive doesn't actually improve the situation. It makes it worse because it makes it harder to notice.

Ordinarily I'd solve this by having a proper message sum-type hierarchy so we can rule out any broken assumptions at compile time but since go doesn't really support sum types we don't have that option. Additionally the switch statement in go isn't capable of informing us of exhaustion, which would be a dramatic improvement over this approach.

Reply:
I'm happy to do this if we can make our test harnesses fail on Critical logs.